File: src/state.py
from dataclasses import dataclass, field
from typing import List, Dict, Any, Optional, Union
import random

@dataclass
class Card:
    id: str
    name: str
    properties: Dict[str, Any]
    owner: Optional[int] = None  # player id if applicable

@dataclass
class Zone:
    name: str
    type: str
    of_deck: Optional[str] = None
    owner: Optional[int] = None      # player id or None for shared
    ordering: Optional[str] = None
    visibility: Optional[Dict[str, str]] = None
    cards: List[Card] = field(default_factory=list)

    @property
    def card_count(self) -> int:
        """Returns the number of cards currently in this zone."""
        return len(self.cards)

    @property
    def top_card(self) -> Optional['Card']:
        """Returns the top card in this zone (last if LIFO/FIFO, first if empty), or None if zone is empty."""
        if not self.cards:
            return None
        return self.cards[-1]

@dataclass
class Player:
    id: int
    name: str
    variables: Dict[str, Any] = field(default_factory=dict)
    zones: Dict[str, Zone] = field(default_factory=dict)

@dataclass
class GameState:
    players: List[Player] = field(default_factory=list)
    shared_zones: Dict[str, Zone] = field(default_factory=dict)
    shared_variables: Dict[str, Any] = field(default_factory=dict)
    decks: Dict[str, List[Card]] = field(default_factory=dict)
    cgml_definition: Any = None     # Optionally reference to loaded CgmlDefinition

def create_deck(deck_def: dict, deck_type_def: Any) -> List[Card]:
    cards = []
    comp = deck_type_def.composition if hasattr(deck_type_def, "composition") else []
    idx = 0
    for entry in comp:
        if entry.get("type") == "template" and entry.get("template") == "standard_suits":
            ranks = entry.get("values")
            suits = ["♠", "♥", "♦", "♣"]
            for suit in suits:
                for rank in ranks:
                    idx += 1
                    card_id = f"{deck_def['type']}-{suit}-{rank}-{idx}"
                    card_name = f"{rank}{suit}"
                    cards.append(Card(
                        id=card_id,
                        name=card_name,
                        properties={"rank": rank, "suit": suit}
                    ))
    return cards

def build_game_state_from_cgml(cgml: Any, player_count: int = None) -> GameState:
    deck_types = cgml.components.component_types.get('deck_types', {}) if cgml.components.component_types else {}
    decks = {}
    for deck_name, deck_def in (cgml.components.decks or {}).items():
        deck_type_def = deck_types.get(deck_def.type, {})
        decks[deck_name] = create_deck(deck_def.dict(), deck_type_def)

    player_count = player_count or cgml.meta.players.max
    players = []
    var_defs = cgml.components.variables if cgml.components.variables else []
    per_player_vars = {v.name: v.initial_value for v in var_defs if v.per_player}
    shared_vars = {v.name: v.initial_value for v in var_defs if not v.per_player}

    zone_defs = cgml.components.zones or []
    shared_zones: Dict[str, Zone] = {}
    per_player_zone_defs = [z for z in zone_defs if getattr(z, "per_player", False)]
    shared_zone_defs = [z for z in zone_defs if not getattr(z, "per_player", False)]

    for pidx in range(player_count):
        pname = f"Player {pidx + 1}"
        player = Player(id=pidx, name=pname)
        player.variables = {k: v for k,v in per_player_vars.items()}
        for zone_def in per_player_zone_defs:
            zone = Zone(
                name=zone_def.name,
                type=zone_def.type,
                of_deck=getattr(zone_def, 'of_deck', None),
                owner=pidx
            )
            player.zones[zone.name] = zone
        players.append(player)

    for zone_def in shared_zone_defs:
        shared_zones[zone_def.name] = Zone(
            name=zone_def.name,
            type=zone_def.type,
            of_deck=getattr(zone_def, 'of_deck', None),
            owner=None
        )

    state = GameState(
        players=players,
        shared_zones=shared_zones,
        shared_variables=shared_vars,
        decks=decks,
        cgml_definition=cgml
    )

    # Assign the cards (by reference!) from deck into the appropriate zone at setup
    for deck_name, cards in decks.items():
        assigned = False
        for zone in state.shared_zones.values():
            if getattr(zone, 'of_deck', None) == deck_name:
                zone.cards.extend(cards)
                assigned = True
        for player in state.players:
            for zone in player.zones.values():
                if getattr(zone, 'of_deck', None) == deck_name:
                    zone.cards.extend(cards)
                    assigned = True
        if not assigned:
            print(f"Warning: Deck '{deck_name}' was generated but not assigned to any zone!")

    return state

def find_zone(state: GameState, zone_path: str, player: Player = None) -> Zone:
    """
    Resolves a dot-path to a Zone object within the GameState.
    - Supports: 'players.0.zones.discard', 'player.1.winnings', 'zones.deck', 'deck'
    - Raises ValueError if the resolved object is not a Zone.

    If player is given and 'zone_path' is a single zone name, looks up in player.zones first.
    """
    # Shortcut: if just a single word, use player zone/then shared
    if '.' not in zone_path:
        if player and zone_path in player.zones:
            return player.zones[zone_path]
        if zone_path in state.shared_zones:
            return state.shared_zones[zone_path]
        for p in state.players:
            if zone_path in p.zones:
                return p.zones[zone_path]
        raise ValueError(f"Zone '{zone_path}' not found.")

    # Build starting context
    ctx = {
        "players": state.players,
        "zones": state.shared_zones,
        "shared_zones": state.shared_zones,
    }
    if player:
        ctx["player"] = player

    current = ctx
    for part in zone_path.split('.'):
        # If list, try integer index
        if isinstance(current, list):
            try:
                idx = int(part)
                current = current[idx]
            except Exception:
                raise KeyError(f"Cannot index list with '{part}'")
        # If dict, use key
        elif isinstance(current, dict):
            if part in current:
                current = current[part]
            else:
                raise KeyError(f"Key '{part}' not found.")
        # If dataclass/object, use attribute
        elif hasattr(current, part):
            current = getattr(current, part)
        else:
            raise KeyError(f"Cannot resolve part '{part}' in object {current}")

    # Final check: ensure this is a Zone
    if not isinstance(current, Zone):
        raise ValueError(f"Path '{zone_path}' does not resolve to a Zone (got {type(current).__name__})")
    return current

def shuffle_zone(zone: Zone):
    random.shuffle(zone.cards)
def move_cards(from_zone: Zone, to_zone: Zone, count: int = 1):
    for _ in range(min(count, len(from_zone.cards))):
        to_zone.cards.append(from_zone.cards.pop(0))

def move_all_cards(from_zone: Zone, to_zone: Zone):
    while from_zone.cards:
        to_zone.cards.append(from_zone.cards.pop(0))

def deal_cards(from_zone: Zone, players: List[Player], to_zone_name: str, count: int):
    for _ in range(count):
        for player in players:
            if from_zone.cards:
                player.zones[to_zone_name].cards.append(from_zone.cards.pop(0))

def deal_all_cards(from_deck: Zone, players: List[Player], to_zone_name: str):
    idx = 0
    pl_count = len(players)
    while from_deck.cards:
        player = players[idx % pl_count]
        player.zones[to_zone_name].cards.append(from_deck.cards.pop(0))
        idx += 1

def perform_setup_action(action: dict, state: GameState):
    """Perform a single setup action as defined in CGML's 'setup'."""
    typ = action['action']
    if typ == "SHUFFLE":
        # Expects 'target': <zone path>
        target = action.get("target")
        if target and target.startswith("zones."):
            # Shuffle per-player or shared
            for player in state.players:
                if target.split('.')[1] in player.zones:
                    shuffle_zone(player.zones[target.split('.')[1]])
            if target.split('.')[1] in state.shared_zones:
                shuffle_zone(state.shared_zones[target.split('.')[1]])
    elif typ == "DEAL":
        from_zone = find_zone(state, action['from'])
        to_zone_name = action['to'].split('.')[-1]  # expects 'zones.hand' style
        count = action['count']
        deal_cards(from_zone, state.players, to_zone_name, count)
    elif typ == "MOVE":
        from_zone = find_zone(state, action['from'])
        to_zone = find_zone(state, action['to'])
        count = action.get('count', 1)
        move_cards(from_zone, to_zone, count)
    elif typ == "MOVE_ALL":
        from_zone = find_zone(state, action['from'])
        to_zone = find_zone(state, action['to'])
        move_all_cards(from_zone, to_zone)
    elif typ == "DEAL_ALL":
        # Used in War (deals all deck cards to per-player zones)
        from_deck_name = action['from_deck']  # e.g., "main_deck"
        to_zone_name = action['to'].split('.')[-1]
        # Find the initial deck zone holding the cards:
        deck_zone = None
        for z in state.shared_zones.values():
            if getattr(z, "of_deck", None) == from_deck_name:
                deck_zone = z
                break
        if not deck_zone:
            for p in state.players:
                for z in p.zones.values():
                    if getattr(z, "of_deck", None) == from_deck_name:
                        deck_zone = z
                        break
        if not deck_zone:
            raise RuntimeError(f"DEAL_ALL: Deck zone for {from_deck_name} not found.")
        deal_all_cards(deck_zone, state.players, to_zone_name)
    else:
        raise NotImplementedError(f"Unknown setup action: {typ}")

def run_setup_phase(state: GameState):
    """Runs all setup actions from the CGML file."""
    for action in state.cgml_definition.setup:
        # If using Pydantic models for actions:
        if hasattr(action, "dict"):
            d = action.dict(by_alias=True)
        else:
            d = dict(action)
        perform_setup_action(d, state)

# --- Usage for setup phase ---

# cgml = load_cgml_file("mygame.yml")
# state = build_game_state_from_cgml(cgml)
# run_setup_phase(state)

File: ts/src/state.ts
export class Card {
    id: string;
    name: string;
    properties: Record<string, any>;
    owner: number | null = null; // player id if applicable

    constructor(id: string, name: string, properties: Record<string, any>, owner: number | null = null) {
        this.id = id;
        this.name = name;
        this.properties = properties;
        this.owner = owner;
    }
}

export class Zone {
    name: string;
    type: string;
    of_deck: string | null = null;
    owner: number | null = null; // player id or null for shared
    ordering: string | null = null;
    visibility: Record<string, string> | null = null;
    cards: Card[] = [];

    constructor(name: string, type: string, of_deck: string | null = null, owner: number | null = null) {
        this.name = name;
        this.type = type;
        this.of_deck = of_deck;
        this.owner = owner;
    }

    get cardCount(): number {
        return this.cards.length;
    }

    get topCard(): Card | null {
        if (this.cards.length === 0) return null;
        return this.cards[this.cards.length - 1];
    }
}

export class Player {
    id: number;
    name: string;
    variables: Record<string, any> = {};
    zones: Record<string, Zone> = {};

    constructor(id: number, name: string) {
        this.id = id;
        this.name = name;
    }
}

export class GameState {
    players: Player[] = [];
    shared_zones: Record<string, Zone> = {};
    shared_variables: Record<string, any> = {};
    decks: Record<string, Card[]> = {};
    cgml_definition: any = null;
    current_state: string | null = null;

    constructor(init?: Partial<GameState>) {
        if (init) Object.assign(this, init);
    }
}

export function createDeck(deckDef: any, deckTypeDef: any): Card[] {
    const cards: Card[] = [];
    const comp = deckTypeDef.composition || [];
    let idx = 0;

    for (const entry of comp) {
        if (entry.type === "template" && entry.template === "standard_suits") {
            const ranks = entry.values;
            const suits = ["♠", "♥", "♦", "♣"];
            for (const suit of suits) {
                for (const rank of ranks) {
                    idx++;
                    const cardId = `${deckDef.type}-${suit}-${rank}-${idx}`;
                    const cardName = `${rank}${suit}`;
                    cards.push(new Card(cardId, cardName, { rank, suit }));
                }
            }
        }
    }
    return cards;
}

export function buildGameStateFromCgml(cgml: any, playerCountOverride?: number): GameState {
    const deckTypes = cgml.components?.component_types?.deck_types || {};
    const decks: Record<string, Card[]> = {};

    const rawDecks = cgml.decks || cgml.components?.decks || {};
    for (const [deckName, deckDef] of Object.entries(rawDecks)) {
        const deckTypeDef = deckTypes[(deckDef as any).type] || {};
        decks[deckName] = createDeck(deckDef, deckTypeDef);
    }

    const playerCount = playerCountOverride || cgml.meta.players.max;
    const players: Player[] = [];

    const varDefs = cgml.variables || cgml.components?.variables || [];
    const perPlayerVars: Record<string, any> = {};
    const sharedVars: Record<string, any> = {};

    for (const v of varDefs) {
        if (v.per_player) perPlayerVars[v.name] = v.initial_value;
        else sharedVars[v.name] = v.initial_value;
    }

    const zoneDefs = cgml.zones || cgml.components?.zones || [];
    const sharedZones: Record<string, Zone> = {};

    const perPlayerZoneDefs = zoneDefs.filter((z: any) => z.per_player);
    const sharedZoneDefs = zoneDefs.filter((z: any) => !z.per_player);

    for (let pidx = 0; pidx < playerCount; pidx++) {
        const pname = `Player ${pidx + 1}`;
        const player = new Player(pidx, pname);
        player.variables = { ...perPlayerVars };
        for (const zoneDef of perPlayerZoneDefs) {
            const zone = new Zone(zoneDef.name, zoneDef.type, zoneDef.of_deck || null, pidx);
            player.zones[zone.name] = zone;
        }
        players.push(player);
    }

    for (const zoneDef of sharedZoneDefs) {
        sharedZones[zoneDef.name] = new Zone(zoneDef.name, zoneDef.type, zoneDef.of_deck || null, null);
    }

    const state = new GameState({
        players,
        shared_zones: sharedZones,
        shared_variables: sharedVars,
        decks,
        cgml_definition: cgml,
    });

    // Assign cards from deck
    for (const [deckName, cards] of Object.entries(decks)) {
        let assigned = false;
        for (const zone of Object.values(state.shared_zones)) {
            if (zone.of_deck === deckName) {
                zone.cards.push(...cards);
                assigned = true;
            }
        }
        for (const player of state.players) {
            for (const zone of Object.values(player.zones)) {
                if (zone.of_deck === deckName) {
                    zone.cards.push(...cards);
                    assigned = true;
                }
            }
        }
        if (!assigned) {
            console.warn(`Warning: Deck '${deckName}' was generated but not assigned to any zone!`);
        }
    }

    return state;
}

export function findZone(state: GameState, zonePath: string | Zone, playerContext?: Player): Zone {
    if (zonePath instanceof Zone) return zonePath;

    // Shortcut
    if (!zonePath.includes('.') && !zonePath.includes('[')) {
        if (playerContext && playerContext.zones[zonePath]) return playerContext.zones[zonePath];
        if (state.shared_zones[zonePath]) return state.shared_zones[zonePath];
        for (const p of state.players) {
            if (p.zones[zonePath]) return p.zones[zonePath];
        }
        throw new Error(`Zone '${zonePath}' not found.`);
    }

    // JSONPath-like start
    if (zonePath.startsWith('$.')) {
        let current: any = {
            players: state.players,
            zones: state.shared_zones,
            shared_zones: state.shared_zones,
        };
        const parts: string[] = [];
        let buf = '';
        for (let i = 2; i < zonePath.length; i++) {
            const ch = zonePath[i];
            if (ch === '.' && !buf.includes('[') && !buf.includes(']')) {
                if (buf) { parts.push(buf); buf = ''; }
                continue;
            }
            buf += ch;
        }
        if (buf) parts.push(buf);

        for (const part of parts) {
            let key = part;
            let idx: number | null = null;
            if (part.includes('[') && part.endsWith(']')) {
                key = part.substring(0, part.indexOf('['));
                const inside = part.substring(part.indexOf('[') + 1, part.length - 1);
                if (inside !== '*') idx = parseInt(inside, 10);
            }
            if (key) {
                current = current[key];
            }
            if (idx !== null) {
                current = current[idx];
            }
        }
        const isObj = current && typeof current === 'object';
        const hasCardsArr = isObj && Array.isArray(current.cards);
        if (isObj && hasCardsArr) return current;
        console.error("findZone duck typing failed! isObj:", isObj, "hasCardsArr:", hasCardsArr, "current.cards type:", isObj ? typeof current.cards : 'N/A', "is Array?", isObj ? Array.isArray(current.cards) : false);
        console.error("findZone failed! current:", current, "Is array?", Array.isArray(current), "Keys?", current ? Object.keys(current) : "null");
        console.error("state.shared_zones:", Object.keys(state.shared_zones));
        throw new Error(`Path '${zonePath}' does not resolve to a Zone`);
    }

    // Dotted paths context
    const ctx: any = {
        players: state.players,
        zones: state.shared_zones,
        shared_zones: state.shared_zones,
    };
    if (playerContext) ctx.player = playerContext;

    let current = ctx;
    for (const part of zonePath.split('.')) {
        if (Array.isArray(current)) {
            const idx = parseInt(part, 10);
            current = current[idx];
        } else if (typeof current === 'object' && current !== null) {
            current = current[part];
        } else {
            throw new Error(`Cannot resolve part '${part}'`);
        }
    }

    if (!(current instanceof Zone)) {
        throw new Error(`Path '${zonePath}' does not resolve to a Zone`);
    }
    return current;
}

export function shuffleZone(zone: Zone): void {
    for (let i = zone.cards.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [zone.cards[i], zone.cards[j]] = [zone.cards[j], zone.cards[i]];
    }
}

export function moveCards(fromZone: Zone, toZone: Zone, count: number = 1): void {
    const amount = Math.min(count, fromZone.cards.length);
    for (let i = 0; i < amount; i++) {
        toZone.cards.push(fromZone.cards.pop() as Card);
    }
}

export function moveAllCards(fromZone: Zone, toZone: Zone): void {
    while (fromZone.cards.length > 0) {
        toZone.cards.push(fromZone.cards.pop() as Card);
    }
}

export function dealCards(fromZone: Zone, players: Player[], toZoneName: string, count: number): void {
    for (let i = 0; i < count; i++) {
        for (const player of players) {
            if (fromZone.cards.length > 0) {
                player.zones[toZoneName].cards.push(fromZone.cards.pop() as Card);
            }
        }
    }
}

export function dealAllCards(fromDeck: Zone, players: Player[], toZoneName: string): void {
    let idx = 0;
    const plCount = players.length;
    while (fromDeck.cards.length > 0) {
        const player = players[idx % plCount];
        player.zones[toZoneName].cards.push(fromDeck.cards.pop() as Card);
        idx++;
    }
}

export function findCardZone(state: GameState, card: Card): Zone | null {
    for (const zone of Object.values(state.shared_zones)) {
        for (const c of zone.cards) {
            if (c.id === card.id) return zone;
        }
    }
    for (const p of state.players) {
        for (const zone of Object.values(p.zones)) {
            for (const c of zone.cards) {
                if (c.id === card.id) return zone;
            }
        }
    }
    return null;
}

export function performSetupAction(action: any, state: GameState): void {
    const typ = action.action;
    if (typ === "SHUFFLE") {
        let target = action.target;
        if (typeof target === 'object' && target.path) target = target.path;

        if (typeof target === 'string') {
            if (target.startsWith("zones.")) {
                const name = target.split('.')[1];
                if (state.shared_zones[name]) shuffleZone(state.shared_zones[name]);
            } else {
                shuffleZone(findZone(state, target));
            }
        }
    } else if (typ === "DEAL") {
        const fromPath = typeof action.from === 'object' ? action.from.path : action.from;
        const toPath = typeof action.to === 'object' ? action.to.path : action.to;
        const count = action.count || 1;
        const fromZone = findZone(state, fromPath);
        const toZoneName = toPath.split('.').pop() as string;
        dealCards(fromZone, state.players, toZoneName, count);
    } else if (typ === "MOVE") {
        const fromPath = typeof action.from === 'object' ? action.from.path : action.from;
        const toPath = typeof action.to === 'object' ? action.to.path : action.to;
        const count = action.count || 1;
        moveCards(findZone(state, fromPath), findZone(state, toPath), count);
    } else if (typ === "MOVE_ALL") {
        const fromPath = typeof action.from === 'object' ? action.from.path : action.from;
        const toPath = typeof action.to === 'object' ? action.to.path : action.to;
        moveAllCards(findZone(state, fromPath), findZone(state, toPath));
    } else if (typ === "DEAL_ALL") {
        const fromPath = typeof action.from === 'object' ? action.from.path : action.from;
        const toPath = typeof action.to === 'object' ? action.to.path : action.to;
        const deckZone = findZone(state, fromPath);
        const toZoneName = toPath.split('.').pop() as string;
        dealAllCards(deckZone, state.players, toZoneName);
    } else {
        throw new Error(`Unknown setup action: ${typ}`);
    }
}

export function runSetupPhase(state: GameState): void {
    for (const action of state.cgml_definition.setup || []) {
        performSetupAction(action, state);
    }
}
